HW2/P3/mandelbrot.pyx

```diff
@@ -12,35 +12,68 @@ cdef np.float64_t magnitude_squared(np.complex64_t z) nogil:
 @cython.boundscheck(False)
 @cython.wraparound(False)
 cpdef mandelbrot(np.complex64_t [:, :] in_coords,
-                 np.uint32_t [:, :] out_counts,
+                 np.float32_t [:, :] out_counts,
                  int max_iterations=511):
+    #np.uint32_t [:, :] out_counts,
     cdef:
        int i, j, iter
+       int width = in_coords.shape[1]
+       int height = in_coords.shape[0]
        np.complex64_t c, z
-
-       # To declare AVX.float8 variables, use:
-       # cdef:
-       #     AVX.float8 v1, v2, v3
-       #
-       # And then, for example, to multiply them
-       #     v3 = AVX.mul(v1, v2)
-       #
-       # You may find the numpy.real() and numpy.imag() fuctions helpful.
+       AVX.float8 c_r, c_i, real_part, img_part, counter_avx, mask_avx, criteria_avx, increment_avx, z_i, z_r, mask_avx2
+       float [:, :] real, img
 
     assert in_coords.shape[1] % 8 == 0, "Input array must have 8N columns"
     assert in_coords.shape[0] == out_counts.shape[0], "Input and output arrays must be the same size"
     assert in_coords.shape[1] == out_counts.shape[1],  "Input and output arrays must be the same size"
 
+    real = numpy.real(in_coords)
+    img = numpy.imag(in_coords)
+
     with nogil:
-        for i in range(in_coords.shape[0]):
-            for j in range(in_coords.shape[1]):
-                c = in_coords[i, j]
-                z = 0
+        for i in prange(height, schedule='static', chunksize=1, num_threads=1):
+            for j in range(width/8):
+
+                # segmenting the values into AVX data structure, real and imaginary parts seperately.
+                z_r = AVX.float_to_float8(0.0)
+                z_i = AVX.float_to_float8(0.0)
+                c_r = AVX.make_float8(real[i, j*8],
+                                           real[i, j*8+1],
+                                           real[i, j*8+2],
+                                           real[i, j*8+3],
+                                           real[i, j*8+4],
+                                           real[i, j*8+5],
+                                           real[i, j*8+6],
+                                           real[i, j*8+7])
+                c_i = AVX.make_float8(img[i, j*8],
+                                          img[i, j*8+1],
+                                          img[i, j*8+2],
+                                          img[i, j*8+3],
+                                          img[i, j*8+4],
+                                          img[i, j*8+5],
+                                          img[i, j*8+6],
+                                          img[i, j*8+7])
```

Comment:
This would have been a clean place to deal with the AVX order reversal since you typed out the entries anyway. Actually, I don't see you handling it before the to_mem() call. Are you writing things to memory backwards?